### ISSUE TITLE
Pressure Vessel silently fails after UID change

### ISSUE THREAD

**Issue by danisztls** [ValveSoftware/steam-runtime#463]
#### Your system information

* Steam Runtime Version: 
* Distribution: Arch Linux
* Link to your full system information: https://gist.github.com/danisztls/15db2ff81be0648cb824c72adb68358f
* Have you checked for system updates?: Yes
* What compatibility tool are you using?: Linux Runtime and Proton Experimental.

* If you are using Steam Linux Runtime for native Linux games: What versions are listed in `SteamLinuxRuntime/VERSIONS.txt`?

`depot	0.20210818.35
LD_LIBRARY_PATH	-	scout	-
scripts	v0.20210818.0-0-gc6be089`

* If you are using Steam Linux Runtime, or Proton 5.13 or newer: What versions are listed in 

`SteamLinuxRuntime_soldier/VERSIONS.txt`?`
depot	0.20210818.24
pressure-vessel	0.20210809.1			
scripts	v0.20210818.0-0-gc6be089
soldier	0.20210817.0	soldier	0.20210817.0
`

#### Please describe your issue in as much detail as possible:

After I changed my UID all non-native games regardless of using LinuxRuntime or Proton failed to open and silently exited. Newly installed games show the same behavior.

`(pressure-vessel-adverb:45904): GLib-WARNING **: getpwuid_r(): failed due to unknown user id (60084)`

[Console STDOUT](https://gist.github.com/danisztls/1cd2e18b50b92f4a7a1054df0716eef5)

#### Steps for reproducing this issue:
1. Install games. 
2. Change user UID.
3. Attempt to play.

**Comment by smcv** [comment]
> Change user UID

Please expand on what you mean by this, in enough detail that someone else could replicate your situation on a test system.

Did you edit system configuration files like `/etc/passwd` to change the numeric uid associated with your username? If yes, did you make sure to change the ownership of all files on disk owned by the old uid to match the new uid, terminate all processes running as the old uid, and start a new login session with the new uid?

Or, do you mean you used something like `sudo` or `pkexec` or some similar tool to get a subtree of processes with a different uid?

**Comment by smcv** [comment]
> `getpwuid_r(): failed due to unknown user id (60084)`

This warning is not *necessarily* fatal, and I'm surprised it wasn't followed by an error message. Please [capture a log](https://github.com/ValveSoftware/steam-runtime/blob/master/doc/reporting-steamlinuxruntime-bugs.md#essential-information) that will give more detail about what is happening.

> `Game 1337010 created interface STEAMAPPLIST_INTERFACE_VERSION001 / AppList`

This looks as though the game (or at least something that loads `libsteam_api.so`) has successfully started in the container, so we must have reached the point at which pressure-vessel hands over control to the game process.

**Comment by smcv** [comment]
> ```
>     Steam Runtime Version:  <Runtime disabled>
> ```

Disabling the Steam Runtime is not supported. It shouldn't have much practical effect on pressure-vessel, but it can break Steam and/or non-pressure-vessel games.

> ```
> The runtime information tool is preparing a report, please wait...
> ```

That isn't a system information report, it's the placeholder that appears before the report has been captured...

**Comment by danisztls** [comment]
> > Change user UID
> 
> Please expand on what you mean by this, in enough detail that someone else could replicate your situation on a test system.
> 
> Did you edit system configuration files like `/etc/passwd` to change the numeric uid associated with your username? If yes, did you make sure to change the ownership of all files on disk owned by the old uid to match the new uid, terminate all processes running as the old uid, and start a new login session with the new uid?
> 
> Or, do you mean you used something like `sudo` or `pkexec` or some similar tool to get a subtree of processes with a different uid?

I'm using **systemd-homed** which does not use `/etc/passwd` and `/etc/shadow` but allows me to encrypt my home directory with **fscrypt** and have improved portability of my user environment between systems. I also I store all my games in a library at `/home/steam`. It isn't a new user as the path suggest but just a non-encrypted directory.

Problem is that I decided to create a new user for a family member to play without thinking it through. As the UID of the new Unix user conflicted with that of my current non-Unix user Homed just irreversibly (AFAIK) bumped it to a random number between 60001 and 60513. That shouldn't be a problem because Homed fix the files ownership when the home directory is changed  but it left Steam Runtime & Proton games not working as well as Gnome Shell extensions broken. I suspect they might be storing and expecting the old UID but I'm also not 100% sure this is it as a recklessly probably messed with the library by running `chmod -R 777 /home/steam`. 

> > getpwuid_r(): failed due to unknown user id (60084)

> This warning is not necessarily fatal, and I'm surprised it wasn't followed by an error message. Please capture a log that will give more detail about what is happening.

> Game 1337010 created interface STEAMAPPLIST_INTERFACE_VERSION001 / AppList

> This looks as though the game (or at least something that loads libsteam_api.so) has successfully started in the container, so we must have reached the point at which pressure-vessel hands over control to the game process.

The error message above disappear when I use the flags bellow and I haven't noticed any additional error message with it on.

```shell
export STEAM_LINUX_RUNTIME_LOG=1
export PRESSURE_VESSEL_VERBOSE=1
```

> Disabling the Steam Runtime is not supported. It shouldn't have much practical effect on pressure-vessel, but it can break Steam and/or non-pressure-vessel games.

I generally use **steam-native** but in this case I did test with **steam-runtime** and results were similar.

Long story short. I removed everything at `.local/share/Steam` and the problem persisted. Then I removed it again but also removed any other Steam resource that I found like `~/.steam` and deleted Runtime and Proton dirs at the Library folder and removed the `compatdata` dir. That solved the problem. :)

**Comment by smcv** [comment]
> As the UID of the new Unix user conflicted with that of my current non-Unix user Homed just irreversibly (AFAIK) bumped it to a random number between 60001 and 60513. That shouldn't be a problem because Homed fix the files ownership when the home directory is changed but it left Steam Runtime & Proton games not working

I don't know how systemd-homed would (know that it has to) fix the ownership of files outside the home directory. If your Steam library is in `/home/steam`, which is outside your systemd-homed-managed home directory `/home/lobo`, then I could well believe that files in `/home/steam` would continue to be owned by your old uid. That is not a configuration that will be compatible with pressure-vessel, which needs to run as the uid that owns `SteamLinuxRuntime_soldier` so that it can create hard-links and fix permissions (Steam's "Steampipe" content-distribution mechanism doesn't preserve permissions, so we have to handle those programmatically).

I would personally recommend allocating a static uid (for example via `homectl --uid`) for each user that is important to you, and certainly allocating a static uid for each user that will own files outside their own home directory. I can see why systemd-homed wants to use dynamic uids, but dynamic mapping between uid and username isn't really consistent with the way Unix filesystems were designed (they store a numeric uid in each inode), so you're going to run into problems whenever you go outside the more "controlled" use-case that systemd-homed manages.

It sounds as though you have worked around this by removing all your Steam data and letting it be re-created, so if there was a bug somewhere in pressure-vessel that is related to this, we will never know what it was - which means there's probably nothing more we can investigate in this issue. If so, we should close it.

> probably messed with the library by running chmod -R 777 /home/steam

I would strongly recommend not doing that. Setting world-writeable permissions on programs that you are going to run (like the games in your Steam library) means that any uid - not just human users' accounts, but also system users that are meant to have absolutely no privileges - can gain control over your account, by altering the game you are about to run to include code of their choice.

If system security is important enough to you that you are encrypting your home directory, then you should probably also be considering other attack vectors like that one. Storing your Steam library with the same level of encryption and DAC security as your home directory (for example, inside your home directory) would be one way to avoid that; running Steam under a different user account would be another way to get privilege-separation.

(I personally use LUKS whole-disk encryption rather than encrypting home directories individually, run Steam under a separate user account so that games don't get access to personal files, and use "fast user switching" to jump between personal and gaming user accounts; but my security/convenience trade-offs don't necessarily match yours.)

**Comment by danisztls** [comment]
> I don't know how systemd-homed would (know that it has to) fix the ownership of files outside the home directory.

I think it only fix those under the home dir but I changed the ownership of those at `/home/steam`.

> I would personally recommend allocating a static uid (for example via homectl --uid) for each user that is important to you, and certainly allocating a static uid for each user that will own files outside their own home directory.

The UID is static and is only change when there's a conflict. In this case it conflicted  with a standard user that I created but home dirs under homed are meant to be portable so anyone can plug a pendrive/disk containing a home dir on any system with the homed daemon and log in.

> isn't really consistent with the way Unix filesystems were designed

Indeed. I'm using homed only to workaround issues that systemd throw when using fscrypt without homed.

> It sounds as though you have worked around this by removing all your Steam data and letting it be re-created, so if there was a bug somewhere in pressure-vessel that is related to this, we will never know what it was - which means there's probably nothing more we can investigate in this issue. If so, we should close it.

Sure! And thanks for the help and the discussion.

> I would strongly recommend not doing that. Setting world-writeable permissions on programs that you are going to run.

It was a lazy attempt fix. :) Using a user group would be better.

> running Steam under a different user account would be another way to get privilege-separation.

Indeed I'm thinking on doing that or tinkering with sandboxing.

> I personally use LUKS whole-disk encryption rather than encrypting home directories individually

I use LUKS/dm-crypt on my HDD but I was with a foot back on using it on a SSD. I don't have enough knowledge to make a proper judgement but there are caveats with that approach on a SSD due to TRIM and wear leveling algos which would affect performance and durability of the drive and could, probably only theoretically, compromise the encryption. The advantage of fscrypt is that encryption is transparent to the filesystem and to the drive.
